benchmark/C/Savina/src/micro/Throughput.lf

```diff
@@ -0,0 +1,183 @@
+// numWorkers = 60
+// [[[end]]]
+
+target C {
+    flags: "-lm",
+    fast: true
+};
+
+
+reactor ThroughputReactor(totalMessages:int(10000)) {
+    
+    preamble {=
+        #include <math.h>
+
+        void performComputation(double theta) {
+            double sint = sin(theta);
+            double res = sint * sint;
+            // defeat dead code elimination
+            if(res <= 0) {
+                printf("Benchmark exited with unrealistic res value.");
+            }
+        }
+    =}
+    
+    input inReset:bool;
+    input inMessage:bool;
+    output outFinished:bool;
+    
+    state messagesProcessed:int(0);
+    
+    reaction(inReset) {=
+        self->messagesProcessed = 0;
+    =}
+    
+    reaction(inMessage) -> outFinished {=
+        self->messagesProcessed++;
+        performComputation(37.2);
+        
+        if(self->messagesProcessed == self->totalMessages) {
+            SET(outFinished, true);
+        }
+    =}
+    
+}
+
+reactor ThroughputProducer(totalMessages:int(10000), numConsumer:int(60)) {
+    
+    state messagesSent:int(0);
+    state nextConsumer:int(0);
+    state numConsumerFinished:int(0);
+    
+    input inStart:bool;
+    output outFinished:bool;
+  
+     /* [[[cog
+      cog.outl(f'output[{numWorkers}] outMessage:bool;')
+      cog.outl(f'input[{numWorkers}] inFinished:bool;')
+      cog.outl(f'output[{numWorkers}] outReset:bool;')
+    ]]] */
+    output[60] outMessage:bool;
+    input[60] inFinished:bool;
+    output[60] outReset:bool;
+    /// [[[end]]]
+    
+    logical action sendNextMessage;
+    
+    reaction(inStart) -> outReset, sendNextMessage {=
+        
+        // reset local state
+        self->messagesSent = 0;
+        self->nextConsumer = 0;
+        self->numConsumerFinished = 0;
+        for(int i = 0; i < self->numConsumer; i++) {
+            SET(outReset[i], true);
+        }
+        
+        // start execution
+        schedule(sendNextMessage, 0);
+    =}
+    
+    reaction(sendNextMessage) -> outMessage, sendNextMessage {=
+        
+        if(self->messagesSent == self->totalMessages) {
+            return;
+        }
+
+        for(int i = 0; i < self->numConsumer; i++) {
+            SET(outMessage[i], true);
+        }
+
+        self->messagesSent += 1;
+        
+        if(self->messagesSent < self->totalMessages) {
+            schedule(sendNextMessage, 0);
+        } 
+    =}
+    
+    reaction(inFinished) -> outFinished {=
+        
+        for(int i = 0; i < self->numConsumer; i++) {
+            if(inFinished[i]->is_present) {
+                self->numConsumerFinished++;
+            }
+        }
+        if(self->numConsumerFinished == self->numConsumer) {
+            SET(outFinished, true);
+        }
+    =}
+}
+
+
+main reactor (numIterations:int(12), numMessagesPerReactor:int(10000)) {
+    
+    /* [[[cog
+      cog.outl(f'producer = new ThroughputProducer(totalMessages=numMessagesPerReactor, numConsumer={numWorkers});')
```

Comment:
Also here, you can make numConsumer a parameter to the main reactor. Then you can pass along the parameter and only need to generate a single line from cog.

Reply:
Do you mean numWorkers?  Or does it not matter?